Nav2SLAMExampleProject/Assets/Scripts/Tests/PlayMode/WaypointIntegrationTest.cs

```diff
@@ -0,0 +1,170 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using NUnit.Framework;
+using Unity.Robotics.Core;
+using Unity.Robotics.ROSTCPConnector;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+using Unity.Robotics.ROSTCPConnector.MessageGeneration;
+using Unity.Robotics.ROSTCPConnector.ROSGeometry;
+
+namespace IntegrationTests 
+{
+    class WaypointTracker
+    {
+        internal Transform CurrentWaypoint => m_Waypoints[m_CurrentWaypointIdx];
+        internal int Count => m_Waypoints.Count;
+        
+        const string k_WaypointTag = "Waypoint"; 
+        List<Transform> m_Waypoints;
+        int m_CurrentWaypointIdx;
+
+        internal WaypointTracker()
+        {
+            var waypoints = GameObject.FindGameObjectsWithTag(k_WaypointTag).ToList();
+            waypoints.Sort((g, o) => string.Compare(g.name, o.name));
+            m_Waypoints = waypoints.Select(w => w.transform).ToList();
+            m_CurrentWaypointIdx = -1;
+            if (m_Waypoints.Count == 0)
+            {
+                Debug.LogWarning(
+                    $"Found no GameObjects tagged with {k_WaypointTag} in {SceneManager.GetActiveScene().name}");
+            }
+        }
+
+        internal bool NextWaypoint()
+        {
+            m_CurrentWaypointIdx++;
+            return m_CurrentWaypointIdx < m_Waypoints.Count;
+        }
+    }
+
+    [TestFixture, Explicit, Category("IntegrationTests")]
+    public class NavigationIntegrationTests
+    {
+        // We don't use Path.Combine here because asset paths are always defined with forward-slash
+        const string k_TestSceneAssetPath = "Scenes/Test/";
+        static readonly List<string> k_TestSceneNames = new List<string>
+        {
+            "SimpleObstacleCourse"
+        };
+
+        // TODO: Use a "TestParameters" MonoBehaviour to hold these parameters in a GameObject per-scene,
+        //       so they can be tuned without re-compiling code
+        const string k_RobotTag = "robot";
+
+        const string k_RobotBaseName = "base_footprint/base_link";
+        const string k_GoalPoseFrameId = "map";
+        const string k_GoalPoseTopic = "/goal_pose";
+        static readonly string k_GoalPoseMessageName = 
+            MessageRegistry.GetRosMessageName<RosMessageTypes.Geometry.PoseStampedMsg>();
+        
+        const float k_Nav2InitializeTime = 5.0f;
+        const float k_SleepBetweenWaypointsTime = 2.0f;
+        
+        // Used to define a timeout for waypoint navigation based on distances between steps
+        const float k_MinimumSpeedExpected = 0.15f;
+
+        // How close the TurtleBot must get to the navigation target to be successful
+        const float k_DistanceSuccessThreshold = 0.4f;
```

Comment:
nit: 0.4f is a bit large compared to the size of the robot?

Reply:
Probably, but I can't tighten it further without the test failing intermittently.  It might be a useful experiment in the future to see how we can get nav2 to more accurately navigate to the goals, but for now this is roughly the minimum threshold to guarantee the test will pass every time.